### DISCUSSION TITLE
Accessing Blob Containers

### DISCUSSION THREAD

**Discussion by DreamDelerium** [Azure/Azurite#2252] [Q&A] [answered]
Hello. I am trying to access files in my Blob Containers but I keep getting 403 errors.  I am trying to run this in Unity, so I am not able to use the Azure SDK's.  Using MS Azure Storage Explorer, I can see the containers and files exist and I can confirm the Primary Key.   This is the code I have:

```
internal class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Hello, World!");
        string authorization, date, apiVersion, fileName;
        Blobs(out authorization, out date, out apiVersion, out fileName);

        using (WebClient client = new WebClient())
        {
            client.Headers.Add("Authorization", $"SharedKey devstoreaccount1:{authorization}");
            client.Headers.Add("x-ms-date", date);
            client.Headers.Add("x-ms-version", apiVersion);

            var s = "http://127.0.0.1:10000/devstoreaccount1/azurite/MyFile.obj";
            client.DownloadString(s);
        }
    }
    static void Blobs(out string auth, out string date, out string apiversion, out string blob)
    {
        string Account = "devstoreaccount1";
        string Key = "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==";
        string Container = "azurite";
        blob = "MyFile.obj";
        apiversion = "2023-08-03";

        DateTime dt = DateTime.UtcNow;
        string StringToSign = String.Format("GET\n"
            + "\n" // content encoding
            + "\n" // content language
            + "\n" // content length
            + "\n" // content md5
            + "\n" // content type
            + "\n" // date
            + "\n" // if modified since
            + "\n" // if match
            + "\n" // if none match
            + "\n" // if unmodified since
            + "\n" // range
            + "x-ms-date:" + dt.ToString("R") + "\nx-ms-version:" + apiversion + "\n" // headers
            + "/{0}/{1}/{2}", Account, Container, blob);

        string signature = SignThis(StringToSign, Key, Account);

        auth = string.Format(
            CultureInfo.InvariantCulture,
            "{0} {1}:{2}",
            "SharedKey",
            Account,
            signature);

        date = dt.ToString("R");
    }

    static String SignThis(String StringToSign, string Key, string Account)
    {
        String signature = string.Empty;
        byte[] unicodeKey = Convert.FromBase64String(Key);
        using (HMACSHA256 hmacSha256 = new HMACSHA256(unicodeKey))
        {
            Byte[] dataToHmac = System.Text.Encoding.UTF8.GetBytes(StringToSign);
            signature = Convert.ToBase64String(hmacSha256.ComputeHash(dataToHmac));
        }

        return signature;
    }
}
```

**Comment by blueww** [comment]
@DreamDelerium 

Since MS Azure Storage Explorer works, it means Azurite as server side works well.  So this looks not an Azurite question, but a question on client side. So our team might not be the best person to answer it.
Since the error is 403, it looks the  StringToSign might be wrong, or the way to get signature from StringToSign  is wrong.

Would you like to get Azurite debug log ([details](https://github.com/Azure/Azurite#debug-log-configuration)), and you can find the StringToSign Azurite used from debug log, and compare with the one your code get.
And you can find the Azurite code to get StringToSign [here](https://github.com/Azure/Azurite/blob/ef4ea8eb5f7130ceb3ea46aa1966e54a19194daa/src/blob/authentication/BlobSharedKeyAuthenticator.ts#L77), and to get signature [here](https://github.com/Azure/Azurite/blob/ef4ea8eb5f7130ceb3ea46aa1966e54a19194daa/src/common/utils/utils.ts#L70).


Besides that, have you tested if your client code works with product Azure storage? If not work, please make sure your code works with public Azure first.

**Comment by blueww** [comment, marked as answer]
I tried to run you code, and find 2problems:

1. the StringToSign is wrong, account name should appear twice in it for emulator, see [details ](https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#:~:text=If%20you%20are%20authorizing%20against%20the%20storage%20emulator%2C%20the%20account%20name%20will%20appear%20twice%20in%20the%20CanonicalizedResource%20string.)
2. You have added ""SharedKey devstoreaccount1:" 2 times in Authorization  header, which makes the header format is wrong.

To fix this, 2 lines of your code need change:
1. `            + "/{0}/{1}/{2}", Account, Container, blob);` should change to `            + "/{0}/{0}/{1}/{2}", Account, Container, blob);`
2. `client.Headers.Add("Authorization", $"SharedKey devstoreaccount1:{authorization}");` should change to `client.Headers.Add("Authorization", authorization);`

**Reply by DreamDelerium** [reply]
Thank you so much!  This worked!
